File: recruitment-server/recruitment-dao/src/main/java/com/andy/recruitment/dao/user/dao/UserDAO.java
package com.andy.recruitment.dao.user.dao;

import com.andy.recruitment.common.user.constant.UserStatus;
import com.andy.recruitment.dao.user.entity.UserInfoDO;
import com.andy.recruitment.dao.user.entity.UserQuery;
import com.andy.spring.page.PageResult;
import com.andy.spring.page.Paginator;
import java.util.List;

/**
 * 用户dao接口
 *
 * @author 庞先海 2020-01-26
 */
public interface UserDAO {

    /**
     * 注册用户信息
     *
     * @param userInfoDo 用户信息
     * @param operator   操作人
     * @return 用户ID
     */
    Long registerUser(UserInfoDO userInfoDo, String operator);

    /**
     * 更新用户信息
     *
     * @param userInfoDo 用户信息
     * @param operator   操作人
     */
    void updateUserInfo(UserInfoDO userInfoDo, String operator);

    /**
     * 更新用户状态
     *
     * @param userId   用户ID
     * @param status   更新后的状态
     * @param operator 操作人
     */
    void updateUserStatus(Long userId, UserStatus status, String operator);

    /**
     * 通过用户ID查询用户信息
     *
     * @param userId 用户ID
     * @return 用户信息
     */
    UserInfoDO getUserInfoByUserId(Long userId);

    /**
     * 通过OpenID查询用户信息
     *
     * @param openId 微信OPENID
     * @return 用户信息
     */
    UserInfoDO getUserInfoByOpenId(String openId);

    /**
     * 通过手机号查询用户
     *
     * @param phone 手机号码
     * @return 用户信息
     */
    UserInfoDO getUserInfoByPhone(String phone);

    /**
     * 分页用户查询
     *
     * @param queryParam 查询参数
     * @param paginator  分页参数
     * @return 分页查询结果
     */
    PageResult<UserInfoDO> getUserInfo(UserQuery queryParam, Paginator paginator);

    /**
     * 通过用户ID列表查询用户ID
     *
     * @param userIdList 用户已ID列表
     * @return 用户信息列表
     */
    List<UserInfoDO> getUserInfo(List<Long> userIdList);

    /**
     * 删除用户信息
     *
     * @param userId 用户ID
     */
    void delete(Long userId);
}

File: recruitment-go/biz/dao/UserDAO.go
package dao

import (
	"errors"
	"recruitment/common"
	"recruitment/constant"
	"recruitment/util"
	"time"
)

type UserInfoDO struct {
	common.BaseDO
	OpenId   string
	UnionId  string
	Nickname string
	Phone    string
	RealName string
	Gender   byte
	Password string
}

func (UserInfoDO) TableName() string {
	return "user_info"
}

type UserInfoQuery struct {
	common.BaseQuery
	ID           uint   `form:"userId" `
	OpenId       string `form:"openId" `
	UnionId      string `form:"unionId" `
	Phone        string `form:"phone" `
	PhoneLike    string `form:"phoneLike" `
	RealNameLike string `form:"realNameLike" `
	Gender       byte   `form:"gender" `
	Status       byte   `form:"status" `
}

type UserDAO struct {
	conn util.DbConnection
}

const (
	userColumns = "id,open_id,union_id,nickname,real_name,gender,phone,password,status," +
		"created_by,created_time,updated_by,updated_time"
)

func (this *UserDAO) GetUserInfoByPhone(phone string) *UserInfoDO {
	var userInfoDO UserInfoDO
	this.conn.GetDb().Select(userColumns).Where("phone = ?", phone).First(&userInfoDO)
	return &userInfoDO
}

func (this *UserDAO) InsertUser(do *UserInfoDO, operator string) (uint, error) {
	do.CreatedTime = time.Now()
	do.CreatedBy = operator
	do.Status = constant.STATUS_NORMAL
	//由于 string 不能赋值为nil 但 open_id 和 union_id 不能重复
	//所以 在 open_id 和 union_id 为空的情况用手机号替代
	if do.OpenId == "" {
		do.OpenId = do.Phone
	}
	if do.UnionId == "" {
		do.UnionId = do.Phone
	}
	db := this.conn.GetDb()
	db = db.Create(do)
	if db.Error != nil {
		return common.USER_ADD_FAILED, db.Error
	}
	return common.SUCCESS, nil
}

func (this *UserDAO) UpdateUser(do *UserInfoDO, operator string) (uint, error) {
	if do.ID == 0 {
		return common.USER_ID_EMPTY, errors.New("user id empty")
	}
	do.UpdatedBy = operator
	do.UpdatedTime = time.Now()
	db := this.conn.GetDb()
	db = db.Model(&do).Omit("id", "created_by", "created_time").Updates(do)
	if db.Error != nil {
		return common.USER_UPDATE_FAILED, db.Error
	}
	return common.SUCCESS, nil
}
